gfdatasource/gfdatasource

```diff
@@ -59,33 +88,136 @@ class GrafanaAPI(object):
             auth=(self.credentials.username, self.credentials.password)
         )
 
+    def update_app(self, app):
+        return requests.post(
+            '/'.join([self.base_url, 'plugins', app.id, 'settings']),
+            json=app.to_json_dict(),
+            auth=(self.credentials.username, self.credentials.password)
+        )
+
+
+def cmd_datasource(grafana_api, opts):
+    datasource_url, datasource_creds = _split_creds(opts.data_source_url)
+    datasource = DataSource(
+        name=opts.name, type=opts.type, access=opts.access,
+        url=datasource_url, credentials=datasource_creds,
+    )
+
+    def update_datasource():
+        grafana_api.update_datasource(datasource)
+
+    return update_datasource
+
+
+def cmd_app(grafana_api, opts):
+    json_data = json.loads(opts.json_data)
+    if opts.secure_json_data is not None:
+        secure_json_data = json.loads(opts.secure_json_data)
+    else:
+        secure_json_data = None
+
+    app = App(
+        id=opts.id, enabled=not opts.disabled, json_data=json_data,
+        secure_json_data=secure_json_data,
+    )
+
+    def update_app():
+        grafana_api.update_app(app)
+
+    return update_app
+
+
+class DefaultSubcommandArgParse(argparse.ArgumentParser):
```

Comment:
Previously the command parser understood `<cmd> <*options>`
Now it understands `<cmd> <global opts> [datasource|app] <specific opts>`
This is here to make `<cmd> <global opts> <datasource opts>` default to the original behaviour.

I have now spotted that it'll do a poor job of separating an interleaving of global and datasource options.